draft-ietf-quic-tls.md

```diff
@@ -1567,12 +1573,13 @@ QUIC requires that the cryptographic handshake provide authenticated protocol
 negotiation.  TLS uses Application Layer Protocol Negotiation (ALPN)
 {{!ALPN=RFC7301}} to select an application protocol.  Unless another mechanism
 is used for agreeing on an application protocol, endpoints MUST use ALPN for
-this purpose.  When using ALPN, endpoints MUST immediately close a connection
-(see Section 10.3 in {{QUIC-TRANSPORT}}) if an application protocol is not
-negotiated with a no_application_protocol TLS alert (QUIC error code 0x178, see
-{{tls-errors}}).  While {{!ALPN}} only specifies that servers use this alert,
-QUIC clients MUST also use it to terminate a connection when ALPN negotiation
-fails.
+this purpose.
+
+When using ALPN, endpoints MUST immediately close a connection (see Section
+10.3 of {{QUIC-TRANSPORT}}) if an application protocol is not negotiated with a
+no_application_protocol TLS alert (QUIC error code 0x178, see {{tls-errors}}).
+While {{!ALPN}} only specifies that servers use this alert, QUIC clients MUST
+use error 0x178 to terminate a connection when ALPN negotiation fails.
```

Comment:
Am I correct to read that this MUST requires a client that receives a no_application_protocol alert to respond with a no_application_protocol alert?

I am not sure if that aligns to what we have now. At the moment, the provisions that we have is that an endpoint that receives CONNECTION_CLOSE MAY respond with a CONNECTION_CLOSE, using NO_ERROR if appropriate (https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-10.3-7).

Based on that, I think what we need here is MAY rather than MUST, and think that NO_ERROR is more appropriate (assuming that we do not have a principle of an endpoint that receives a CONNECTION_CLOSE frame carrying an error echo the error code it has received). Or it might make sense to just drop the last sentence.

Reply:
I think this is intended to apply to the case when a server does not include an ALPN extension.  If the client receives an alert, ALPN isn't really the thing that failed -- the handshake is.